test/paper_input_test.dart

```diff
@@ -0,0 +1,194 @@
+        expect(input.noLabelFloat, isFalse);
+        var floatingLabel = Polymer
+            .dom(Polymer
+                .dom(input.jsElement['root'])
+                .querySelector('paper-input-container').jsElement['root'])
+            .querySelector('.label-is-floating');
+        expect(floatingLabel, isNotNull);
+      });
+
+      test('always-float-label attribute works without placeholder', () {
+        var input = fixture('always-float-label');
+        var container = Polymer
+            .dom(input.jsElement['root'])
+            .querySelector('paper-input-container');
+        var inputContent = Polymer
+            .dom(container.jsElement['root'])
+            .querySelector('.input-content');
+        expect(inputContent.classes.contains('label-is-floating'), isTrue);
+      });
+
+      test('error message is displayed', () {
+        var input = fixture('error');
+        forceXIfStamp(input);
+        var error = Polymer
+            .dom(input.jsElement['root'])
+            .querySelector('paper-input-error');
+        expect(error, isNotNull);
+        expect(error.getComputedStyle().display, isNot(equals('none')));
+      });
+
+      test('empty required input shows error', () {
+        var input = fixture('required');
+        forceXIfStamp(input);
+        var error = Polymer
+            .dom(input.jsElement['root'])
+            .querySelector('paper-input-error');
+        expect(error, isNotNull);
+        expect(error.getComputedStyle().display, isNot(equals('none')));
+      });
+
+      test('character counter is displayed', () {
+        var input = fixture('char-counter');
+        forceXIfStamp(input);
+        PaperInputCharCounter counter = Polymer
+            .dom(input.jsElement['root'])
+            .querySelector('paper-input-char-counter');
+        expect(counter, isNotNull);
+        expect(
+            counter.jsElement['_charCounterStr'], equals(input.value.length));
+      });
+
+      test('validator is used', () {
+        var input = fixture('validator');
+        expect(input.inputElement.invalid, isTrue);
+      });
+
+      test('caret position is preserved', () {
+        var input = fixture('basic');
+        var ironInput = Polymer
+            .dom(input.jsElement['root'])
+            .querySelector('input[is="iron-input"]');
+        input.value = 'nananana';
+        ironInput.selectionStart = 2;
+        ironInput.selectionEnd = 2;
+        input.updateValueAndPreserveCaret('nanananabatman');
+        expect(ironInput.selectionStart, equals(2));
+        expect(ironInput.selectionEnd, equals(2));
+      });
+    });
+
+    group('focus/blur events', () {
+      PaperInput input;
+      setUp(() {
+        input = fixture('basic');
+      });
+      test('focus/blur events fired on host element', () {
+        var nFocusEvents = 0;
+        var nBlurEvents = 0;
+        input.on['focus'].listen((event) {
```

Comment:
Before each of these `.listen` calls its good to add a `.take(1)` call, assuming the event is only going to get called once for the test. That way the stream automatically cancels itself after the first event.